### DISCUSSION TITLE
Problem with No MEM-AP AHB3/5 found

### DISCUSSION THREAD

**Discussion by J-Wrobel** [majbthrd/DapperMime#10] [Q&A] [answered]
Hello, 
I loaded DapperMime on one rp2040 and I am trying to debug a second rp2040. I compiled OpenOCD following mainly instructions in pico_debug repo. During connection I get error:

`Debug: 100 1633 arm_adi_v5.c:952 dap_find_ap(): No MEM-AP AHB3 found`
`Debug: 101 2969 arm_adi_v5.c:952 dap_find_ap(): No MEM-AP AHB5 found`

I have loaded on the debugged rp2040 blink by file transfer. After I try to debug it with OpenOCD blinking stops until hard reset. It seems this error prevents me from loading binaries onto rp2040 with the use of DAP or even connecting to gdb (halt error). I am new to debugging with DAP and OpenOCD. Anyone encountered such thing or can give me some pointers were to look for cause? I checked sources of OpenOCD but could not find any clue as to where to look further. 

Log in attachment.
[debug memo.txt](https://github.com/majbthrd/DapperMime/files/8446592/debug.memo.txt)

**Comment by majbthrd** [comment, marked as answer]
@J-Wrobel, first thank you for posting in Discussions.  So many lazy people create a code Issue even though there is not a code issue.

Alas, it is not viable to borrow instructions for a different project and expect the same results.

You should not use the pico-debug instructions for DapperMime.

pico-debug is special in that it will work with both authentic OpenOCD and raspberrypi/openocd.

What you are trying to do was debug an entire RP2040 Pico chip, and at the moment, this is only possible with raspberrypi/openocd.

The RP2040 Pico is unusual in that debugging the chip (by connecting to the SWD connector) depends on something called "SWD multi-drop".  At the time of writing, the authentic OpenOCD still doesn't have all the functionality needed for this.  It does work with pico-debug, which doesn't need "SWD multi-drop".

pico-debug is different in that it runs on Core1 and debugs Core0; it makes Core0 look like the only CPU.  What you were attempting was to connect to both Core0 and Core1, which needs "SWD multi-drop"... even if you are only debugging one core.

For DapperMime, what I suggest is that you do is follow the instructions in "Appendix A: Using Picoprobe", but make the following substitutions.  Here is the document:

https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf

Follow "Build OpenOCD", but instead of:

```
./configure --enable-picoprobe
```

do:

```
./configure --enable-cmsis-dap
```

Note that you *MUST* have source code from raspberrypi/openocd that is no older than 2021 Jun 18 since raspberrypi stalled on adopting my CMSIS-DAP patch until then.

Or, if you don't want to build it yourself (particular if you are not using Linux), you can get pre-built binaries here:

https://github.com/earlephilhower/pico-quick-toolchain/releases/

Then follow "Picoprobe Wiring".

Then, skip "Install Picoprobe driver (only needed on Windows)", since DapperMime doesn't need drivers.

Then, for "Using Picoprobe with OpenOCD", instead of:

```
src/openocd -f interface/picoprobe.cfg -f target/rp2040.cfg -s tcl
```

do:

```
src/openocd -f interface/cmsis-dap.cfg -c "adapter speed 4000" -f target/rp2040.cfg -s tcl
```

The OpenOCD output should look something like this:

```
Open On-Chip Debugger 0.11.0-g610f137-dirty (2022-04-10-11:37)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
adapter speed: 4000 kHz

Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 4000 kHz
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x00000001
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x10000001
Info : rp2040.core0: hardware has 4 breakpoints, 2 watchpoints
Info : rp2040.core1: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for rp2040.core0 on 3333
Info : Listening on port 3333 for gdb connections
```

So, the executive summary is that the only changes from the picoprobe instructions are:

**a) no special drivers needed**
**b) build OpenOCD with "--enable-cmsis-dap" instead of (or in addition to) "--enable-picoprobe"**
**c) run OpenOCD with:**
```
	-f interface/cmsis-dap.cfg -c "adapter speed 4000"
```
   **instead of:**
```
	-f interface/picoprobe.cfg
```

You should be able to use any of the many instructions online that are written for picoprobe just with the replacements (b) and (c) above.

**Comment by J-Wrobel** [comment]
Thanks for quick reply. I figured it would be 'me' issue rather than 'code' issue :). You are right that it was quite naive for me to expect success not following exact instructions on topic that is fairly new to me. I guess I felt over-adventurous. I followed most of instructions from DapperMime but did build regular OpenOCD miss-understanding that lacking support for multi-drop would limit me to one core. I need to do some more reading on that but You are probably right that it actually prevents me to access any core at all. Thanks again for taking time to provide instructions and some insight on multi-drop.
